File: backend/schemas/simulation.py
from __future__ import annotations

from pydantic import BaseModel, Field, model_validator
from typing import Literal


class SimulationRequest(BaseModel):
    """
    Parameters controlling the Monte Carlo market shock simulation.
    All values have sensible defaults; clients may override any field.
    """

    # Simulation scale
    n_steps: int = Field(
        252, ge=50, le=1000,
        description="Simulation horizon in trading days. 252 ≈ 1 trading year.",
    )
    n_runs: int = Field(
        200, ge=10, le=2000,
        description="Number of Monte Carlo paths per strategy.",
    )
    n_agents: int = Field(
        5, ge=2, le=20,
        description="Number of correlated market participants whose noise is Cholesky-decomposed.",
    )

    # Market parameters
    baseline_volatility: float = Field(
        0.15, ge=0.01, le=0.80,
        description="Annualized baseline return volatility (pre-shock).",
    )
    market_sensitivity: float = Field(
        0.85, ge=0.1, le=2.0,
        description="Portfolio universe sensitivity to the systemic market factor.",
    )

    # Shock parameters
    shock_type: Literal["interest_rate", "liquidity", "volatility", "contagion"] = Field(
        "liquidity",
        description="Category of macroeconomic shock to inject.",
    )
    shock_intensity: float = Field(
        0.4, ge=0.0, le=1.0,
        description="Shock magnitude (0 = no shock, 1 = maximum distress).",
    )
    shock_start: int = Field(
        80, ge=5,
        description="Simulation step at which the shock begins propagating.",
    )

    # Dynamics
    recovery_speed: float = Field(
        0.08, ge=0.005, le=0.5,
        description="Exponential decay rate lambda of shock effects. Higher = faster recovery.",
    )
    contagion_strength: float = Field(
        0.5, ge=0.0, le=1.0,
        description="Inter-agent correlation amplification during shock periods.",
    )

    # Adaptive strategy
    adaptive_aggressiveness: float = Field(
        0.7, ge=0.05, le=3.0,
        description="UCB1 exploration constant C. Higher values favour arm exploration.",
    )

    @model_validator(mode="after")
    def shock_start_valid(self) -> SimulationRequest:
        if self.shock_start >= self.n_steps - 10:
            raise ValueError(
                f"shock_start ({self.shock_start}) must be at least 10 steps "
                f"before n_steps ({self.n_steps})."
            )
        return self


class MetricsResult(BaseModel):
    """Quantitative performance and risk metrics for a single strategy."""
    avg_return: float = Field(description="Mean terminal return across all Monte Carlo runs.")
    terminal_value: float = Field(description="Mean terminal portfolio value (initial = 1.0).")
    volatility: float = Field(description="Annualized realized volatility of daily returns.")
    max_drawdown: float = Field(description="Maximum peak-to-trough drawdown on the mean path (negative).")
    recovery_time: int = Field(description="Steps from shock onset to recovery of pre-shock level.")
    resilience_score: float = Field(description="Composite resilience score in [0, 1].")
    sharpe_ratio: float = Field(description="Annualized Sharpe ratio (risk-free = 2.5% p.a.).")
    sortino_ratio: float = Field(description="Sortino ratio using downside deviation.")


class StrategyResult(BaseModel):
    """Complete simulation output for one strategy."""
    name: str
    mean_path: list[float] = Field(description="Mean portfolio value at each time step.")
    percentile_10: list[float] = Field(description="10th-percentile path across Monte Carlo runs.")
    percentile_90: list[float] = Field(description="90th-percentile path across Monte Carlo runs.")
    metrics: MetricsResult


class SimulationResponse(BaseModel):
    """Full response from /api/simulate."""
    strategies: list[StrategyResult]
    shock_start: int
    n_steps: int
    best_by_resilience: str = Field(description="Strategy with highest resilience_score.")
    best_by_return: str = Field(description="Strategy with highest avg_return.")
    best_by_sharpe: str = Field(description="Strategy with highest Sharpe ratio.")
    scenario_label: str = Field(description="Human-readable shock scenario label.")

File: frontend/types/simulation.ts
export type ShockType =
  | "interest_rate"
  | "liquidity"
  | "volatility"
  | "contagion";

export interface SimulationParams {
  n_steps: number;
  n_runs: number;
  n_agents: number;
  baseline_volatility: number;
  shock_intensity: number;
  market_sensitivity: number;
  recovery_speed: number;
  contagion_strength: number;
  adaptive_aggressiveness: number;
  shock_start: number;
  shock_type: ShockType;
}

export interface MetricsResult {
  avg_return: number;
  terminal_value: number;
  volatility: number;
  max_drawdown: number;
  recovery_time: number;
  resilience_score: number;
  sharpe_ratio: number;
  sortino_ratio: number;
}

export interface StrategyResult {
  name: string;
  mean_path: number[];
  percentile_10: number[];
  percentile_90: number[];
  metrics: MetricsResult;
}

export interface SimulationResponse {
  strategies: StrategyResult[];
  shock_start: number;
  n_steps: number;
  best_by_resilience: string;
  best_by_return: string;
  best_by_sharpe: string;
  scenario_label: string;
}

export const DEFAULT_PARAMS: SimulationParams = {
  n_steps: 252,
  n_runs: 200,
  n_agents: 5,
  baseline_volatility: 0.15,
  shock_intensity: 0.4,
  market_sensitivity: 0.85,
  recovery_speed: 0.08,
  contagion_strength: 0.5,
  adaptive_aggressiveness: 0.7,
  shock_start: 80,
  shock_type: "liquidity",
};

export const STRATEGY_COLORS: Record<string, string> = {
  "Conservative Hedging": "#3B82F6",
  "Balanced Reallocation": "#10B981",
  "Aggressive Risk-Seeking": "#EF4444",
  "Adaptive Strategy": "#8B5CF6",
};

export const STRATEGY_SHORT: Record<string, string> = {
  "Conservative Hedging": "Conservative",
  "Balanced Reallocation": "Balanced",
  "Aggressive Risk-Seeking": "Aggressive",
  "Adaptive Strategy": "Adaptive",
};

export const SHOCK_LABELS: Record<ShockType, string> = {
  interest_rate: "Interest Rate Shock",
  liquidity: "Liquidity Crisis",
  volatility: "Volatility Spike",
  contagion: "Systemic Contagion",
};

export const SHOCK_DESCRIPTIONS: Record<ShockType, string> = {
  interest_rate:
    "A sudden rise in policy rates increases discount factors, compresses asset valuations, and sustains a persistent drag on portfolio returns.",
  liquidity:
    "Market liquidity dries up, triggering forced selling, wide bid-ask spreads, and sharp volatility amplification across all correlated agents.",
  volatility:
    "A volatility regime shift (VIX-style spike) dramatically increases return variance, creating large dispersion across strategies and runs.",
  contagion:
    "Losses propagate through an inter-agent network: initial losses at one node transmit to correlated counterparties, building to a systemic peak before gradual recovery.",
};
